File: core/server/project/commonSceneBase/src/main/java/com/home/commonSceneBase/scene/scene/BattleSceneSyncPlayLogic.java
package com.home.commonSceneBase.scene.scene;

import com.home.commonSceneBase.scene.scene.BSceneMethodLogic;

/** 帧同步副本逻辑 */
public class BattleSceneSyncPlayLogic extends BSceneMethodLogic
{
	///** 帧计时 */
	//private int _frameTime=0;
	///** 帧序号 */
	//private int _frameIndex=0;
	///** 帧数据 */
	//private SList<FrameSyncData> _frames=new SList<>(FrameSyncData[]::new);
	///** 接收指令组 */
	//private SList<FrameSyncCommandData> _receiveCommands=new SList<>(FrameSyncCommandData[]::new);
	//
	//@Override
	//protected void onStart()
	//{
	//	super.onStart();
	//
	//	for(int i=Global.frameSyncCacheFrames=-1;i>=0;--i)
	//	{
	//		pushFrame(false);
	//	}
	//
	//	_scene.setFrameSync(true);
	//
	//	_scene.aoi.radioMessageAll(FrameSyncStartRequest.create(_frames));
	//}
	//
	//@Override
	//public void onFrame(int delay)
	//{
	//	super.onFrame(delay);
	//
	//	//还未开始
	//	if(!_scene.isFrameSync())
	//		return;
	//
	//	int t=_frameTime+delay;
	//
	//	while(t>= ShineSetting.systemFrameDelay)
	//	{
	//		t-=ShineSetting.systemFrameDelay;
	//		pushFrame(true);
	//	}
	//
	//	_frameTime=t;
	//}
	//
	//private void pushFrame(boolean needSend)
	//{
	//	FrameSyncData data=new FrameSyncData();
	//	data.index=_frameIndex++;
	//
	//	SList<FrameSyncCommandData> receiveCommands;
	//
	//	if(!(receiveCommands=_receiveCommands).isEmpty())
	//	{
	//		data.commands=new SList<>(FrameSyncCommandData[]::new,receiveCommands.size());
	//
	//		receiveCommands.forEachAndClear(v->
	//		{
	//			data.commands.add(v);
	//		});
	//	}
	//
	//	_frames.add(data);
	//
	//	if(needSend)
	//	{
	//		_scene.aoi.radioMessageAll(FrameSyncFrameRequest.create(data));
	//	}
	//}
	//
	//public void receiveCommand(long playerID,FrameSyncCommandData data)
	//{
	//	//还未开始
	//	if(!_scene.isFrameSync())
	//		return;
	//
	//	//写序号
	//	data.index=_scene.getCharacterByPlayerID(playerID).getUnitData().getCharacterIdentity().syncIndex;
	//
	//	_receiveCommands.add(data);
	//}
}

File: core/client/game/src/commonGame/scene/scene/BattleSceneSyncPlayLogic.cs
using System;
using ShineEngine;

/// <summary>
/// 副本帧同步场景逻辑
/// </summary>
public class BattleSceneSyncPlayLogic:SceneMethodLogic
{
	/** 帧计时 */
	private int _frameTime=0;

	/** 缓冲组 */
	private SQueue<FrameSyncData> _cacheQueue=new SQueue<FrameSyncData>();

	private bool _waiting=false;

	/** 帧同步开始 */
	public void onFrameStart(SList<FrameSyncData> list)
	{
		//去掉头三帧

		list.forEach(k=>
		{
			_cacheQueue.offer(k);
		});

		_scene.setFrameSync(true);
	}

	/** 接收帧 */
	public void receiveFrame(FrameSyncData data)
	{
		_cacheQueue.offer(data);

		if(_waiting)
		{
			if(_cacheQueue.size()>=Global.frameSyncCacheFrames)
			{
				_waiting=false;
			}
		}

		while(_cacheQueue.Count>Global.frameSyncCacheFrames)
		{
			doFrame(ShineSetting.systemFrameDelay);
		}
	}

	/** 预备帧 */
	public void preFrame(int delay)
	{
		int t=_frameTime+delay;

		while(t>=ShineSetting.systemFrameDelay)
		{
			t-=ShineSetting.systemFrameDelay;
			doFrame(ShineSetting.systemFrameDelay);
		}

		_frameTime=t;
	}

	private void doFrame(int delay)
	{
		if(_waiting)
			return;

		if(_cacheQueue.isEmpty())
		{
			_waiting=true;
		}
		else
		{
			FrameSyncData data=_cacheQueue.poll();

			if(data.commands!=null && !data.commands.isEmpty())
			{
				data.commands.forEach(preExecuteCommand);
			}

			_scene.doFrame(delay);
		}
	}

	/** 预备执行指令 */
	protected virtual void preExecuteCommand(FrameSyncCommandData command)
	{
		Unit unit=_scene.getCharacterByIndex(command.index);

		if(unit==null)
		{
			Ctrl.throwError("帧同步出错");
			return;
		}

		executeCommand(unit,command);
	}

	/** 执行指令 */
	protected virtual void executeCommand(Unit unit,FrameSyncCommandData command)
	{

	}
}